README.md
# AirHooperAttack

AirHopper Attack <https://arxiv.org/pdf/1411.0237.pdf> Reproduce


Sender/ChangeLog
2018-11-05  tempest_for_leishen
	* add leishen.cpp for FM boarding

2001-12-01  tempest_for_eliza-1.0.5
	* added mp3 playback :-)

2001-08-14  tempest_for_eliza-1.0.4
	* added include <string.h>

2001-08-14  tempest_for_eliza-1.0.3
	* program now runs fullscreen.

2001-03-28  tempest_for_eliza-1.0.2
	* added a new song

2001-02-17  tempest_for_eliza-1.0.1
	* switched to SDL because most people don't have ggi.

1999-08-24  tempest_for_eliza-1.0.0
	* initial release


Sender/trans.cpp
/*
  tempest for eliza code
  Copyright (C) 2001  Erik Thiele

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <math.h>
#include <stdlib.h>
#include <sys/time.h>
#include <unistd.h>
#include <stdio.h>
#include <SDL.h>
#include <string.h>

double carrier;
int resx;
int resy;
int horizontalspan;
double pixelclock;

SDL_Surface *screen;
Uint32 white, black;

class zeit : public timeval {
public:
  zeit() { setzero(); };
  void getcurrenttime() { gettimeofday (this,0); };
  void subtract (const zeit &subme)
    {
      tv_sec-=subme.tv_sec;
      tv_usec-=subme.tv_usec;
      normalize(); 
    };
  void add (const zeit &addme)
    {
      tv_sec+=addme.tv_sec;
      tv_usec+=addme.tv_usec;
      normalize(); 
    };
  void normalize()
    {
      while (tv_usec<0) {
	tv_usec+=1000000;
	tv_sec--;
      };
      while (tv_usec>=1000000) {
	tv_usec-=1000000;
	tv_sec++;
      };
    };
  bool zero()
    {
      return (tv_sec==0)&&(tv_usec==0);
    };
  void setzero()
    {
      tv_sec=tv_usec=0;
    };
  void print()
    {
      printf("%8d:%8d\n",int(tv_sec),int(tv_usec));
    };
  void setdouble(const double val)
    {
      long long i=(long long)(val*1000000);
      tv_usec=i%1000000;
      tv_sec=i/1000000;
      normalize();
    };
};

inline void pixelchen (const int x, const int y, const Uint32 pixel)
{
  Uint8 *bits=((Uint8*)screen->pixels)+y*screen->pitch+x*screen->format->BytesPerPixel;
  switch(screen->format->BytesPerPixel) {
  case 1:
    *((Uint8 *)(bits)) = (Uint8)pixel;
    break;
  case 2:
    *((Uint16 *)(bits)) = (Uint16)pixel;
    break;
  case 3:
    { /* Format/endian independent */
      Uint8 r, g, b;
      
      r = (pixel>>screen->format->Rshift)&0xFF;
      g = (pixel>>screen->format->Gshift)&0xFF;
      b = (pixel>>screen->format->Bshift)&0xFF;
      *((bits)+screen->format->Rshift/8) = r; 
      *((bits)+screen->format->Gshift/8) = g;
      *((bits)+screen->format->Bshift/8) = b;
    }
    break;
  case 4:
    *((Uint32 *)(bits)) = (Uint32)pixel;
    break;
  };
};

void waituntil (const zeit &desttime)
{
  zeit tm(desttime);
  zeit cur;
  cur.getcurrenttime();
  tm.subtract(cur);

  select(1,0,0,0,&tm);
};

void mkrealsound(double freq)
{
  SDL_Rect myrect;
  myrect.x=myrect.y=0;
  myrect.w=resx;
  myrect.h=resy;
  SDL_FillRect(screen,&myrect,black);

  printf("mkrealsound(): freq=%f\n", freq);

  int x, y;
  double t;
  
/*
  double ftfp2 = freq / pixelclock * 2.0;
  double fcfp2 = carrier / pixelclock * 2.0;
  
  t = 0;
  for (y = 0; y < resy; y++) {
    if (int(t*ftfp2)%2) {
      // 1
      t+=resx;
    }else{
      // 0
      for (x = 0; x < resx; x++) {
	if (int(t*fcfp2)%2)
	  pixelchen(x,y,white);
	t++;
      };
    };
    
    t += horizontalspan - resx;
  }
*/
  double k = 2 * freq / pixelclock;
  double k2 = 2 * carrier / pixelclock;
  t = 0;
  for (y = 0; y < resy; y ++) {
    if (int(t * k) % 2) {
      for (x = 0; x < resx; x ++) {
        if (int(t * k2) % 2) {
          pixelchen(x,y,white);
        }
        t ++;
      }
    } else {
      t = t + resx;
    }
  }
}

double note (int nr)
{
  double n = pow(2.0,nr/12.0)*440.0*2;
  printf("note(): nr=%d, n=%f\n", nr, n);
  return n;
};


int curoct, curnot;
double curlen;
bool loaded;
zeit timefornextsound;

void resetplay()
{
  curoct=0;
  curnot=0;
  curlen=128;
  loaded=false;
  timefornextsound.setzero();
};

void finishplay ()
{
  SDL_Rect myrect;
  myrect.x=myrect.y=0;
  myrect.w=resx;
  myrect.h=resy;
  SDL_FillRect(screen,&myrect,black);

  waituntil(timefornextsound);
  SDL_UpdateRect(screen,0,0,resx,resy);
};

void flushplay ()
{
  if (loaded) {
    printf("flushplay(): curnot=%d,curoct=%d,curlen=%f\n",curnot, curoct, curlen);
    if (curnot==-1000) {
      SDL_Rect myrect;
      myrect.x=myrect.y=0;
      myrect.w=resx;
      myrect.h=resy;
      SDL_FillRect(screen,&myrect,black);

      waituntil(timefornextsound);
    }else{
      mkrealsound(note(curnot+12*curoct));
      //mkrealsound(22000.00f+random()%1000);
    };
    if (timefornextsound.zero()) {
      timefornextsound.getcurrenttime();
    }else{
      waituntil(timefornextsound);
    };
    SDL_UpdateRect(screen,0,0,resx,resy);

    zeit length;
    length.setdouble(curlen/40.0);
    timefornextsound.add(length);
    loaded=false;
  };
};

void play (const char *const song)
{
  int pos=0;
  while (pos<int(strlen(song))) {
    SDL_Event event;
    while(SDL_PollEvent(&event)) if (event.type == SDL_MOUSEBUTTONDOWN) exit(0);
    switch(song[pos]) {
    case 'c': printf("c\n");curnot=0;     loaded=true; flushplay();break;
    case 'd': printf("d\n");curnot=2;     loaded=true; flushplay(); break;
    case 'e': printf("e\n");curnot=4;     loaded=true; flushplay(); break;
    case 'f': printf("f\n");curnot=5;     loaded=true; flushplay(); break;
    case 'g': printf("g\n");curnot=7;     loaded=true; flushplay(); break;
    case 'a': printf("a\n");curnot=9;     loaded=true; flushplay(); break;
    case 'b': printf("b\n");curnot=11;    loaded=true; flushplay(); break;
    case 'p': printf("p\n");curnot=-1000; loaded=true; flushplay(); break;
    case '#': printf("#\n");curnot++;     loaded=true; flushplay(); break;
    case '1': printf("1\n");curlen=128; break;
    case '2': printf("2\n");curlen=64;  break;
    case '4': printf("4\n");curlen=32;  break;
    case '5': printf("5\n");curlen=48;  break;
    case '8': printf("8\n");curlen=16;  break;
    case '6': printf("6\n");curlen=8;   break;
    case 'u': printf("u\n");curoct=-2;  break;
    case 'l': printf("l\n");curoct=-1;  break;
    case 'h': printf("h\n");curoct=0;   break;
    case 'x': printf("x\n");curoct=1;   break;
    default:break;
    };
    pos++;
  };
  flushplay();
};

void usage()
{
  printf("\nwrong parameters ! read readme file!\n\n");
  exit(1);
};

int main(int argc, char *argv[])
{
  carrier=10000000.0;
  resx=1024;
  resy=768;
  horizontalspan=1400;
  pixelclock=105.0 * 1e6;
  char *filename;

  atexit(SDL_Quit);

  printf(
	 "\n"
	 "Tempest for Binge - by yaojie !\n"
	 "--------------------------------\n"
	 "\n"
	 );

  if (argc!=7) usage();
  pixelclock=atof(argv[1]);
  resx=atol(argv[2]);
  resy=atol(argv[3]);
  horizontalspan=atol(argv[4]);
  carrier=atof(argv[5]);
  filename=argv[6];

  printf("\n"
	 "Pixel Clock %.0f Hz\n"
	 "X Resolution %d Pixels\n"
	 "Y Resolution %d Pixels\n"
	 "Horizontal Total %d Pixels\n"
	 "AM Carrier Frequency %.0f Hz\n"
	 "\n\n",
	 pixelclock,resx,resy,horizontalspan,carrier);

  if ( SDL_Init(SDL_INIT_VIDEO) < 0 ) {
    fprintf(stderr, "Couldn't initialize SDL: %s\n", SDL_GetError());
    exit(1);
  };
  
  /* Have a preference for 8-bit, but accept any depth */
  screen = SDL_SetVideoMode(resx, resy, 8, SDL_SWSURFACE|SDL_ANYFORMAT|SDL_FULLSCREEN);
  if ( screen == NULL ) {
    fprintf(stderr, "Couldn't set video mode: %s\n", SDL_GetError());
    exit(1);
  }
  printf("Have Set %d bits-per-pixel mode\n",
	 screen->format->BitsPerPixel);

  SDL_Color col[2];
  col[0].r=col[0].g=col[0].b=0xff; // white
  col[1].r=col[1].g=col[1].b=0x00; // black
  
  SDL_SetColors(screen,col,0,2);
  white = SDL_MapRGB(screen->format,0xff,0xff,0xff);
  black = SDL_MapRGB(screen->format,0x00,0x00,0x00);
  
  char song[100000];
  FILE *input=fopen(filename,"r");
  if (!input) { printf("cannot open file\n"); return 1; };
  resetplay();
  char *ret;
  do {
    ret=fgets(song,100000,input);
    if (ret) {
      for (unsigned i=0;i<strlen(song);i++)
	      if (song[i]=='*') song[i]=0;
      play (song);
    }
  } while(ret);
  finishplay();
  fclose(input);

  return 0;
};


Sender/leishen.cpp
/*
  tempest for Leishen code
  Copyright (C) 2001  Erik Thiele
  Copyright (C) 2018  ITB Yaojie

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <math.h>
#include <stdlib.h>
#include <sys/time.h>
#include <unistd.h>
#include <stdio.h>
#include <SDL.h>
#include <string.h>

double carrier;
int resx;
int resy;
int horizontalspan;
double pixelclock;

SDL_Surface *screen;
Uint32 white, black;

class zeit : public timeval {
public:
  zeit() { setzero(); };
  void getcurrenttime() { gettimeofday (this,0); };
  void subtract (const zeit &subme)
    {
      tv_sec-=subme.tv_sec;
      tv_usec-=subme.tv_usec;
      normalize(); 
    };
  void add (const zeit &addme)
    {
      tv_sec+=addme.tv_sec;
      tv_usec+=addme.tv_usec;
      normalize(); 
    };
  void normalize()
    {
      while (tv_usec<0) {
	tv_usec+=1000000;
	tv_sec--;
      };
      while (tv_usec>=1000000) {
	tv_usec-=1000000;
	tv_sec++;
      };
    };
  bool zero()
    {
      return (tv_sec==0)&&(tv_usec==0);
    };
  void setzero()
    {
      tv_sec=tv_usec=0;
    };
  void print(const char * prefix)
    {
      printf("%s():%8d:%8d\n",prefix,int(tv_sec),int(tv_usec));
    };
  void setdouble(const double val)
    {
      long long i=(long long)(val*1000000);
      tv_usec=i%1000000;
      tv_sec=i/1000000;
      normalize();
    };
};

inline void pixelchen (const int x, const int y, const Uint32 pixel)
{
  Uint8 *bits=((Uint8*)screen->pixels)+y*screen->pitch+x*screen->format->BytesPerPixel;
  switch(screen->format->BytesPerPixel) {
    case 1:
      *((Uint8 *)(bits)) = (Uint8)pixel;
      break;
    case 2:
      *((Uint16 *)(bits)) = (Uint16)pixel;
      break;
    case 3:
      { /* Format/endian independent */
        Uint8 r, g, b;
      
        r = (pixel>>screen->format->Rshift)&0xFF;
        g = (pixel>>screen->format->Gshift)&0xFF;
        b = (pixel>>screen->format->Bshift)&0xFF;
        *((bits)+screen->format->Rshift/8) = r; 
        *((bits)+screen->format->Gshift/8) = g;
        *((bits)+screen->format->Bshift/8) = b;
      }
      break;
    case 4:
      *((Uint32 *)(bits)) = (Uint32)pixel;
      break;
  };
};

void waituntil (const zeit &desttime)
{
  zeit tm(desttime);
  zeit cur;
  cur.getcurrenttime();
  tm.subtract(cur);
  //tm.print("waituntil");

  select(1,0,0,0,&tm);
};

struct DTMF {
  double low;
  double high;
};

void mkrealsound(struct DTMF freq)
{
  SDL_Rect myrect;
  myrect.x=myrect.y=0;
  myrect.w=resx;
  myrect.h=resy;
  SDL_FillRect(screen,&myrect,black);

  //printf("mkrealsound(): freq=%f, %f\n", freq.low, freq.high);

  int x, y;
  double t;
  
  double lk = 2 * freq.low / pixelclock;
  double hk = 2 * freq.high / pixelclock;

  double k2 = 2 * carrier / pixelclock;
  t = 0;
  for (y = 0; y < resy; y ++) {
    if (y < resy / 2) {
      if (int(t * lk) % 2) {
        for (x = 0; x < resx; x ++) {
          if (int(t * k2) % 2) {
            pixelchen(x,y,white);
          }
          t ++;
        }
      } else {
        t = t + resx;
      }
    } else {
      if (int(t * hk) % 2) {
        for (x = 0; x < resx; x ++) {
          if (int(t * k2) % 2) {
            pixelchen(x,y,white);
          }
          t ++;
        }
      } else {
        t = t + resx;
      }
    }
  }
}

void mkrealsound(int nr) {
  SDL_Rect myrect;
  myrect.x=myrect.y=0;
  myrect.w=resx;
  myrect.h=resy;
  SDL_FillRect(screen,&myrect,black);

  int x, y;
  double t;
  double k1, k2;
  printf(" %d ", nr);
  
  if (nr == 0) {
    k1 = 697.0f;
  } else {
    k1 = 1633.0f;
  }

  k1 = 2 * k1 / pixelclock;
  k2 = 2 * carrier / pixelclock;
  t = 0;
  for (y = 0; y < resy; y ++) {
    if (int(t * k1) % 2) {
      for (x = 0; x < resx; x ++) {
        if (int(t * k2) % 2) {
          pixelchen(x,y,white);
        }
        t ++;
      }
    } else {
      t = t + resx;
    }
  }
}

double curlen=10.0f;
zeit timefornextsound;

void resetplay()
{
  timefornextsound.setzero();
};

void flushread() 
{
  SDL_FillRect(screen,0,black);
}

void finishplay ()
{
  flushread();
  waituntil(timefornextsound);
  SDL_UpdateRect(screen,0,0,resx,resy);
};


void setDelayTime() {
  zeit length;
  length.setdouble(curlen/40.0);
  timefornextsound.add(length);
}

void showBlackScreen() {
  flushread();
  SDL_UpdateRect(screen,0,0,0,0);
  setDelayTime();
  waituntil(timefornextsound);
}

void waitawhile() {
  if (timefornextsound.zero()) {
    timefornextsound.getcurrenttime();
  }

  setDelayTime();
  waituntil(timefornextsound);
  showBlackScreen();
}

void read (const char *const book)
{
  int pos=0;
  while (pos<int(strlen(book))) {
    SDL_Event event;
    while(SDL_PollEvent(&event)) if (event.type == SDL_MOUSEBUTTONDOWN) exit(0);
    char x = book[pos];
    printf("\nx=%c : ", x);
    for (int i = 0; i < 8; i ++) {
      mkrealsound((x >> (7-i)) & 0x1);
      SDL_UpdateRect(screen,0,0,resx,resy);
      waitawhile();
    }
    pos++;
  }
  flushread();
};

void read_prefix() {
  const char x[] = {0x11, 0x11, 0x11, 0x11, 0x0};
  read(x);
}

void usage()
{
  printf("\nwrong parameters ! read readme file!\n\n");
  exit(1);
};

int main(int argc, char *argv[])
{
  carrier=10000000.0;
  resx=1024;
  resy=768;
  horizontalspan=1400;
  pixelclock=105.0 * 1e6;
  char *filename;

  atexit(SDL_Quit);

  printf(
	 "\n"
	 "Tempest for LeiShen - by itb !\n"
	 "--------------------------------\n"
	 "\n"
	 );

  if (argc!=7) usage();
  pixelclock=atof(argv[1]);
  resx=atol(argv[2]);
  resy=atol(argv[3]);
  horizontalspan=atol(argv[4]);
  carrier=atof(argv[5]);
  filename=argv[6];

  printf("\n"
	 "Pixel Clock %.0f Hz\n"
	 "X Resolution %d Pixels\n"
	 "Y Resolution %d Pixels\n"
	 "Horizontal Total %d Pixels\n"
	 "AM Carrier Frequency %.0f Hz\n"
	 "\n\n",
	 pixelclock,resx,resy,horizontalspan,carrier);

  if ( SDL_Init(SDL_INIT_VIDEO) < 0 ) {
    fprintf(stderr, "Couldn't initialize SDL: %s\n", SDL_GetError());
    exit(1);
  };
  
  /* Have a preference for 8-bit, but accept any depth */
  screen = SDL_SetVideoMode(resx, resy, 8, SDL_SWSURFACE|SDL_ANYFORMAT|SDL_FULLSCREEN);
  if ( screen == NULL ) {
    fprintf(stderr, "Couldn't set video mode: %s\n", SDL_GetError());
    exit(1);
  }
  printf("Have Set %d bits-per-pixel mode\n",
	 screen->format->BitsPerPixel);

  SDL_Color col[2];
  col[0].r=col[0].g=col[0].b=0xff; // white
  col[1].r=col[1].g=col[1].b=0x00; // black
  
  SDL_SetColors(screen,col,0,2);
  white = SDL_MapRGB(screen->format,0xff,0xff,0xff);
  black = SDL_MapRGB(screen->format,0x00,0x00,0x00);
  printf("main(): black=%d, white=%d\n", black, white);
  
  char book[100000];
  FILE *input=fopen(filename,"r");
  if (!input) { printf("cannot open file\n"); return 1; };
  resetplay();
  char *ret;
  do {
    ret=fgets(book,100000,input);
    if (ret) {
      read_prefix();
      read (book);
    }
  } while(ret);
  finishplay();
  fclose(input);

  return 0;
};


Receiver/decoder.py
import wave
import matplotlib.pyplot as plt
import numpy as np
import os
 
filepath  = "C:/Users/yaojieno1/Music/"
filename  = "20181027-7th-2.wav"
isShowBin = False
threshold = 0
framerate = 0
nframes   = 0
waveData  = []

isDebug = False

def log(fmt):
    if isDebug:
        print(fmt)

def readfile():
    global filepath, filename, framerate, waveData, nframes
    
    # 读取文件，获取通道数/帧率
    f = wave.open(filepath+filename,'rb')
    params = f.getparams()
    nchannels, sampwidth, framerate, nframes = params[:4]
    log("nchannels=" + str(nchannels)+" samwidth="+str(sampwidth)+" framerate="+str(framerate)+" nframes="+str(nframes))

    # 读取音频值
    strData = f.readframes(nframes)
    waveData = np.fromstring(strData,dtype=np.uint8)

    # 归一化
    waveData = waveData * 1.0 / 128  #(max(abs(waveData)))
    waveData = np.reshape(waveData,[nframes,nchannels]).T

    print("len(waveData)=" + str(len(waveData[0])))
    f.close()

#==============

def search_max(arr):
    m = 0
    c = -1

    for i in range(len(arr)):
        if arr[i].real > m:
            m = arr[i].real
            c = i
    return c, m

#==============

def calc_freq(start, end, index):
    global waveData, framerate, threshold

    if end == 0:
        end = start + 1024
    
    N  = end - start       # 采样个数
    df = framerate / (N-1) # 分辨率
    freq = [ df * n for n in range(0, N) ]

    data=waveData[0][start:end]

    c = np.fft.fft(data) # * 2 / N
    e = int(len(c) / 2)
    while freq[e] > 2000:   
        e -= 10

    b = 0
    while freq[b] < 400:
        b += 10

    if index > 0:
        plt.figure(index)
        plt.plot(freq[b:e-1], abs(c[b:e-1]))

    i, m = search_max(abs(c[b:e-1]))
  
    if (i > 0):
        i += b

    log("m=" + str(m) + " i=" + str(i) + " freq[i]=" + str(freq[i])  + " start=" + str(start) + " end=" + str(end))

    if m < threshold or i == -1:
        return -1

    if (freq[i] < 600):
        return 0

    if (freq[i] > 1100):
        return 1

    log("{EXCEPTION] output = 8" + " m=" + str(m) + " i=" + str(i) + " start=" + str(start) + " end=" + str(end) + " freq[i]=" + str(freq[i]))
    return 8


def show_result(final):
    global threshold, isShowBin

    if (len(final) < 1):
        return

    for v in final:
        o = ord(v)
        if (o == 0xA or o == 0x9 or o==0xD or o==0x11):
            continue
        if (o < 0x20 or o > 0x7F):
            if (False == isShowBin):
                return
    print("threshold:" + str(threshold) + " result: " + ("".join(final)))


def proc():
    global nframes

    i = 0
    t = -1
    val = []
    final = []
    while i < nframes:
        s = calc_freq(i, 0, -1)
        if s == -1 or s == 8:
            if t != -1:
                log("t == -1, i=" + str(i))
                t = -1
        elif s == 1:
            if t == 0.7:
                log("val[].append(1), i=" + str(i))
                val.append(1)
                t = 1
            elif t != 1:
                t = 0.7
                log("t=0.7, i=" + str(i));
        elif s == 0:
            if t == 0.2:
                log("val[].append(0), i=" + str(i))
                val.append(0)
                t = 0
            elif t != 0:
                t = 0.2
                log("t=0.2, i=" + str(i));
                

        i += 200
        if(len(val) == 8):
            v = 0
            for j in range(8):
                v = (v << 1) + val[j]
            log("v=" + hex(v) + " chr=" + chr(v))
            final.append(chr(v))
            val = []
    show_result(final)


def check(beg, end):
    n = 1
    k = beg
    while k <= end:
        print(calc_freq(k, 0, n))
        n += 1
        k += 200
    plt.show()

def process():
    global threshold

    for x in range(10,36):
        threshold = x / 10
        proc()


def main():
    readfile()
    process()
    # check(5000, 8000)


main()


Sender/player.cpp
/*
  tempest for eliza code
  Copyright (C) 2001  Erik Thiele

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <math.h>
#include <stdlib.h>
#include <sys/time.h>
#include <unistd.h>
#include <stdio.h>
#include <SDL.h>
#include <string.h>

double carrier;
int resx;
int resy;
int horizontalspan;
double pixelclock;

SDL_Surface *screen;
Uint32 white, black;

class zeit : public timeval {
public:
  zeit() { setzero(); };
  void getcurrenttime() { gettimeofday (this,0); };
  void subtract (const zeit &subme)
    {
      tv_sec-=subme.tv_sec;
      tv_usec-=subme.tv_usec;
      normalize(); 
    };
  void add (const zeit &addme)
    {
      tv_sec+=addme.tv_sec;
      tv_usec+=addme.tv_usec;
      normalize(); 
    };
  void normalize()
    {
      while (tv_usec<0) {
	tv_usec+=1000000;
	tv_sec--;
      };
      while (tv_usec>=1000000) {
	tv_usec-=1000000;
	tv_sec++;
      };
    };
  bool zero()
    {
      return (tv_sec==0)&&(tv_usec==0);
    };
  void setzero()
    {
      tv_sec=tv_usec=0;
    };
  void print()
    {
      printf("%8d:%8d\n",int(tv_sec),int(tv_usec));
    };
  void setdouble(const double val)
    {
      long long i=(long long)(val*1000000);
      tv_usec=i%1000000;
      tv_sec=i/1000000;
      normalize();
    };
};

inline void pixelchen (const int x, const int y, const Uint32 pixel)
{
  Uint8 *bits=((Uint8*)screen->pixels)+y*screen->pitch+x*screen->format->BytesPerPixel;
  switch(screen->format->BytesPerPixel) {
  case 1:
    *((Uint8 *)(bits)) = (Uint8)pixel;
    break;
  case 2:
    *((Uint16 *)(bits)) = (Uint16)pixel;
    break;
  case 3:
    { /* Format/endian independent */
      Uint8 r, g, b;
      
      r = (pixel>>screen->format->Rshift)&0xFF;
      g = (pixel>>screen->format->Gshift)&0xFF;
      b = (pixel>>screen->format->Bshift)&0xFF;
      *((bits)+screen->format->Rshift/8) = r; 
      *((bits)+screen->format->Gshift/8) = g;
      *((bits)+screen->format->Bshift/8) = b;
    }
    break;
  case 4:
    *((Uint32 *)(bits)) = (Uint32)pixel;
    break;
  };
};

void waituntil (const zeit &desttime)
{
  zeit tm(desttime);
  zeit cur;
  cur.getcurrenttime();
  tm.subtract(cur);

  select(1,0,0,0,&tm);
};

void mkrealsound(double freq)
{
  SDL_Rect myrect;
  myrect.x=myrect.y=0;
  myrect.w=resx;
  myrect.h=resy;
  SDL_FillRect(screen,&myrect,black);

  printf("mkrealsound(): freq=%f\n", freq);

  int x, y;
  double t;
  
/*
  double ftfp2 = freq / pixelclock * 2.0;
  double fcfp2 = carrier / pixelclock * 2.0;
  
  t = 0;
  for (y = 0; y < resy; y++) {
    if (int(t*ftfp2)%2) {
      // 1
      t+=resx;
    }else{
      // 0
      for (x = 0; x < resx; x++) {
	if (int(t*fcfp2)%2)
	  pixelchen(x,y,white);
	t++;
      };
    };
    
    t += horizontalspan - resx;
  }
*/
  double k = 2 * freq / pixelclock;
  double k2 = 2 * carrier / pixelclock;
  t = 0;
  for (y = 0; y < resy; y ++) {
    if (int(t * k) % 2) {
      for (x = 0; x < resx; x ++) {
        if (int(t * k2) % 2) {
          pixelchen(x,y,white);
        }
        t ++;
      }
    } else {
      t = t + resx;
    }
  }
}

double note (int nr)
{
  double n = pow(2.0,nr/12.0)*440.0*2;
  printf("note(): nr=%d, n=%f\n", nr, n);
  return n;
};


int curoct, curnot;
double curlen;
bool loaded;
zeit timefornextsound;

void resetplay()
{
  curoct=0;
  curnot=0;
  curlen=128;
  loaded=false;
  timefornextsound.setzero();
};

void finishplay ()
{
  SDL_Rect myrect;
  myrect.x=myrect.y=0;
  myrect.w=resx;
  myrect.h=resy;
  SDL_FillRect(screen,&myrect,black);

  waituntil(timefornextsound);
  SDL_UpdateRect(screen,0,0,resx,resy);
};

void flushplay ()
{
  if (loaded) {
    printf("flushplay(): curnot=%d,curoct=%d,curlen=%f\n",curnot, curoct, curlen);
    if (curnot==-1000) {
      SDL_Rect myrect;
      myrect.x=myrect.y=0;
      myrect.w=resx;
      myrect.h=resy;
      SDL_FillRect(screen,&myrect,black);

      waituntil(timefornextsound);
    }else{
      mkrealsound(note(curnot+12*curoct));
    };
    if (timefornextsound.zero()) {
      timefornextsound.getcurrenttime();
    }else{
      waituntil(timefornextsound);
    };
    SDL_UpdateRect(screen,0,0,resx,resy);

    zeit length;
    length.setdouble(curlen/40.0);
    timefornextsound.add(length);
    loaded=false;
  };
};

void play (const char *const song)
{
  int pos=0;
  while (pos<int(strlen(song))) {
    SDL_Event event;
    while(SDL_PollEvent(&event)) if (event.type == SDL_MOUSEBUTTONDOWN) exit(0);
    switch(song[pos]) {
    case 'c': printf("c\n");flushplay(); curnot=0; loaded=true; break;
    case 'd': printf("d\n");flushplay(); curnot=2; loaded=true; break;
    case 'e': printf("e\n");flushplay(); curnot=4; loaded=true; break;
    case 'f': printf("f\n");flushplay(); curnot=5; loaded=true; break;
    case 'g': printf("g\n");flushplay(); curnot=7; loaded=true; break;
    case 'a': printf("a\n");flushplay(); curnot=9; loaded=true; break;
    case 'b': printf("b\n");flushplay(); curnot=11; loaded=true; break;
    case 'p': printf("p\n");flushplay(); curnot=-1000; loaded=true; break;
    case '#': printf("#\n");curnot++; break;
    case '1': printf("1\n");flushplay(); curlen=128; break;
    case '2': printf("2\n");flushplay(); curlen=64; break;
    case '4': printf("4\n");flushplay(); curlen=32; break;
    case '5': printf("5\n");flushplay(); curlen=48; break;
    case '8': printf("8\n");flushplay(); curlen=16; break;
    case '6': printf("6\n");flushplay(); curlen=8; break;
    case 'u': printf("u\n");flushplay(); curoct=-2; break;
    case 'l': printf("l\n");flushplay(); curoct=-1; break;
    case 'h': printf("h\n");flushplay(); curoct=0; break;
    case 'x': printf("x\n");flushplay(); curoct=1; break;
    default:break;
    };
    pos++;
  };
  flushplay();
};

void usage()
{
  printf("\nwrong parameters ! read readme file!\n\n");
  exit(1);
};

int main(int argc, char *argv[])
{
  carrier=10000000.0;
  resx=1024;
  resy=768;
  horizontalspan=1400;
  pixelclock=105.0 * 1e6;
  char *filename;

  atexit(SDL_Quit);

  printf(
	 "\n"
	 "Tempest for Eliza - by erikyyy !\n"
	 "--------------------------------\n"
	 "\n"
	 "Read the README file to understand what's happening\n"
	 "if you do not read it, you will NOT know what to do\n"
	 );

  if (argc!=7) usage();
  pixelclock=atof(argv[1]);
  resx=atol(argv[2]);
  resy=atol(argv[3]);
  horizontalspan=atol(argv[4]);
  carrier=atof(argv[5]);
  filename=argv[6];

  printf("\n"
	 "Pixel Clock %.0f Hz\n"
	 "X Resolution %d Pixels\n"
	 "Y Resolution %d Pixels\n"
	 "Horizontal Total %d Pixels\n"
	 "AM Carrier Frequency %.0f Hz\n"
	 "\n\n",
	 pixelclock,resx,resy,horizontalspan,carrier);

  if ( SDL_Init(SDL_INIT_VIDEO) < 0 ) {
    fprintf(stderr, "Couldn't initialize SDL: %s\n", SDL_GetError());
    exit(1);
  };
  
  /* Have a preference for 8-bit, but accept any depth */
  screen = SDL_SetVideoMode(resx, resy, 8, SDL_SWSURFACE|SDL_ANYFORMAT|SDL_FULLSCREEN);
  if ( screen == NULL ) {
    fprintf(stderr, "Couldn't set video mode: %s\n", SDL_GetError());
    exit(1);
  }
  printf("Have Set %d bits-per-pixel mode\n",
	 screen->format->BitsPerPixel);

  SDL_Color col[2];
  col[0].r=col[0].g=col[0].b=0xff; // white
  col[1].r=col[1].g=col[1].b=0x00; // black
  
  SDL_SetColors(screen,col,0,2);
  white = SDL_MapRGB(screen->format,0xff,0xff,0xff);
  black = SDL_MapRGB(screen->format,0x00,0x00,0x00);
  
  char song[100000];
  FILE *input=fopen(filename,"r");
  if (!input) { printf("cannot open file\n"); return 1; };
  resetplay();
  char *ret;
  do {
    ret=fgets(song,100000,input);
    if (ret) {
      for (unsigned i=0;i<strlen(song);i++)
	      if (song[i]=='*') song[i]=0;
      play (song);
    }
  } while(ret);
  finishplay();
  fclose(input);

  return 0;
};
